### PR TITLE
Refactor mesh information into exporters

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR refactors the mesh, parallel context, and location into the exporters.
Hence, there is now one `Exporter` instance per mesh.

## Motivation and additional information

This simplifies the participant and the handling of filenames.
It also allows adding `series` files #1991.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.22.1.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/io/ExportCSV.cpp [hunk 1]
```diff
@@ -35,35 +35,41 @@ struct StridedAccess {
...
   namespace fs = std::filesystem;
-  fs::path outfile(location);
-  if (not location.empty()) {
+  fs::path outfile(_location);
+  if (not _location.empty()) {
     fs::create_directories(outfile);
   }
   outfile /= filename;
 
+  const auto &mesh = *_mesh;
```
Comment (davidscn) [inline src/io/ExportCSV.cpp:71]: Why is this necessary? If you work with a pointer before, I would prefer a pointer in the following as well.

File: src/io/ExportVTK.cpp [hunk 2]
```diff
@@ -17,28 +17,38 @@
...
 {
-  PRECICE_TRACE(name, location, mesh.getName());
-  PRECICE_ASSERT(name != std::string(""));
-  PRECICE_ASSERT(!utils::IntraComm::isParallel(), "ExportVTK only supports serial participants.");
+  PRECICE_TRACE(index, time, _mesh->getName());
+  PRECICE_ASSERT(index >= 0);
+  PRECICE_ASSERT(time >= 0.0);
+  PRECICE_ASSERT(!isParallel(), "ExportVTK only supports serial participants.");
+
+  auto filename = fmt::format("{}-{}.{}{}.vtk", _participantName, _mesh->getName(), kindPrefix(), index);
```
Comment (davidscn) [inline src/io/ExportVTK.cpp:37]: This changes the current export names, right? Could affect the users and or the tests.

File: src/io/ExportXML.cpp [hunk 3]
```diff
@@ -138,6 +157,9 @@ void ExportXML::writeSubFile(
   outSubFile << (utils::isMachineBigEndian() ? "BigEndian\">" : "LittleEndian\">") << '\n';
 
   outSubFile << "   <" << formatType << ">\n";
+
+  const auto &mesh = *_mesh;
```
Comment (davidscn) [inline src/io/ExportXML.cpp:161]: Same here

**Review by davidscn — APPROVED** [review]
Looks good, only minor things

File: src/io/ExportVTK.cpp (hunk 2 above)
Comment (fsimonis) [reply src/io/ExportVTK.cpp:37]: Actually, I think this is the only instance, where the filename is the same.

My idea was to make paraview detect all exports as series. I haven't though about backwards compatibility.

I'll revert the name changes.

File: src/io/ExportCSV.cpp (hunk 1 above)
Comment (fsimonis) [reply src/io/ExportCSV.cpp:71]: It prevents having to change the uses of mesh in the remaining code. I'll clean it up.

**Comment by fsimonis** [comment]
@davidscn this may be a good opportunity to change the `inital` export to `dt0`. Would you consider this a bugfix or a breaking change?

**Comment by davidscn** [comment]
> @davidscn this may be a good opportunity to change the `inital` export to `dt0`. Would you consider this a bugfix or a breaking change?

If anything then breaking. It was or is not a bug and should be mostly a debugging feature anyway. It might be, however, that people use the exports in scripts, we also use it in the replay mode for example (also in the systemtests @MakisH ?). Not sure what will happen if we rename it here without clear communication and or backwards compatibility.

**Comment by fsimonis** [comment]
I doubt that the inital exports are used especially as they never contained data up to version 3 #1563. Are they used/loaded in ASTE?

**Comment by davidscn** [comment]
https://github.com/precice/aste/blob/bdc3fbb45051a5f3721d56a390a798b399ca33c3/src/mesh.cpp#L384-L390

**Comment by MakisH** [comment]
We currently do have `.init.vtk` files in the reference results of the tutorials (i.e., system tests), but we can update the reference results to remove those.

**Comment by fsimonis** [comment]
OK, reverting to init

**Comment by fsimonis** [comment]
Patched some edge cases. Merging now

**Comment by MakisH** [comment]
> We currently do have `.init.vtk` files in the reference results of the tutorials (i.e., system tests), but we can update the reference results to remove those.

> OK, reverting to init

Did not react early enough to this: My comment was only an FYI. I was not opposing the `.init`. I think that eventually it would make sense to replace the `init` with `0`, as also in the training course, it appears as a duplicate entry in ParaView (which in that context definitely generates questions).

**Comment by fsimonis** [comment]
> it appears as a duplicate entry in ParaView 

so dt0 == init?

**Comment by MakisH** [comment]
No, I meant that you see the `init` and the group `1..N` separately. But this is why you started this discussion, I guess.

**Comment by fsimonis** [comment]
Ah. That's part of the motivation for #2055. Now there is one file you can load which also exposes the init export as t=0.

**Comment by MakisH** [comment]
We held back the `.init` to not break backwards compatibility, but it looks like we did break it by swapping the order of mesh and participant:

- previously: `Fluid-Mesh-Fluid.dt1.vtk`
- now: `Fluid-Fluid-Mesh.dt1.vtk`

While this makes sense, it now breaks the system tests making debugging a bit complicated (some bugs are actual bugs, but all tests fail because of the file renaming).

Could we revert this change until we can do the breaking in a more controlled and documented way?
